build: update dev-infra package to latest version with full ESM

devversion opened the pull request:
See individual commits

devversion committed:
ci: update renovate to update `angular` packages when available

We recently aggressively limited Renovate to only two days and set the
hourly limit to one. This resulted in e.g. Angular not being updated
for quite a while now, as well as missing out on e.g. dev-infra updates
which should happen as often as possible.

This commit reworks the config to (1) avoid legacy syntax and (2) to
update Angular-cross repo depsas often as possible. Also groups Bazel
dependencies better and attempts to fix the non-working Babel dep
grouping.

devversion committed:
build: update `dev-infra` package to latest version with full ESM

Updates `angular/dev-infra` to the latest version which is now full ESM.

devversion committed:
build: wire up `ng-dev` ts-node config loading

With the new ESM version of `ng-dev`, the TypeScript config loading
cannot happen automatically because `ts-node` cannot hook into the
Node ESM module resolution at runtime. To fix this `ng-dev` is now
loading a plain `.ng-dev/config.mjs` and the runtime just needs to be
configured accordingly, i.e. running `ng-dev` with `ts-node` so that
the ESM loader can be wired up before Node's runtime starts (using e.g.
`node --loader`).

alan-agius4 commented on renovate.json:
```diff
@@ -37,67 +41,46 @@
-      ],
-      "matchPackagePatterns": ["*"],
-      "matchUpdateTypes": ["minor", "patch"],
-      "matchCurrentVersion": ">=1",
```
`"matchCurrentVersion": ">=1",` is still needed as otherwise Renovate would update the placeholder versions. 

Example
https://github.com/angular/angular/blob/f94c6f433dba3924b79f137cfcc49d2dfd4d679c/packages/common/package.json#L23

devversion replied:
Good point, I wasn't sure what this is for. Would it be better to have another section where we just match directly `0.0.0-PLACEHOLDER`? I wanted to also update versions smaller than `1.0.0`.

alan-agius4 replied:
Probably we can use the not syntax. 

```
``

devversion committed:
build: replace copied `spec_bundle` rule in devtools with dev-infra rule

Replaces the copied `spec_bundle` rule in the devtools folder with the
one provided by dev-infra. Initially we couldn't use the rule because
it wasn't available / and also later didn't work with the linker
version from `compiler-cli` built from `HEAD`.

alan-agius4 commented on renovate.json:
```diff
@@ -1,17 +1,21 @@
-  "semanticCommits": true,
-  "semanticPrefix": "build",
+  "pinDigests": true,
+  "rebaseWhen": "behind-base-branch",
```
I don't think we should not enable this option.

This was enabled in https://github.com/angular/ngcc-validation/blob/master/renovate.json and created a lot of unnecessary rebasing. 

Also now, rebasing would also cause approvals to be lost.

IMHO, `behind-base-branch` should only be considered if renovate manages PR merges, which of course we don't want to.

devversion replied:
Good point on the post approval check

devversion replied:
without the behind base branch option, renovate won't rebase PRs if they are starting to conflict right?

alan-agius4 commented on renovate.json:
```diff
@@ -37,67 +41,51 @@
     {
-      "packageNames": ["typescript", "rxjs", "tslib"],
-      "updateTypes": ["minor", "major"],
+      "matchPackageNames": ["typescript", "rxjs", "tslib"],
```
This changes the behaviour of renovate as now updates to these packages is disabled, but previously we did allow minor and patch updates.

```js
  {
      "packageNames": ["typescript", "rxjs", "tslib"],
      "separateMinorPatch": true
    },
    {
      "packageNames": ["typescript", "rxjs", "tslib"],
      "updateTypes": ["minor", "major"],
      "enabled": false
    },
```

devversion replied:
Fixed as discussed offline. We will just make TS and tslib a group and let renovate creates PRs for updates of these rarely changing deps

alan-agius4 commented on renovate.json:
```diff
@@ -37,67 +41,51 @@
-      "matchPackagePatterns": ["*"],
-      "groupName": "scorecard action dependencies",
-      "groupSlug": "scorecard-action"
+      "matchCurrentVersion": "0.0.0-PLACEHOLDER",
```
👍

alan-agius4 commented on renovate.json:
```diff
@@ -37,67 +41,51 @@
-      "groupName": "bazel",
-      "pinVersions": false
+      "matchPackagePrefixes": ["@angular/", "angular/", "@angular-devkit", "@schematics/"],
+      "followTag": "next",
```
Question: do we want AIO to always use the `next` version of the CLI? I was under the impression that we don't.

devversion replied:
Yeah, I think we always would want to stay/test as up-to-date as possible. Especially since we deploy AIO from `HEAD`/`main` or from RC/patch branches. We also did this manually in this past and this would ease the process: https://github.com/angular/angular/commits/main/aio/package.json

alan-agius4 reviewed (changes requested):
Couple of another comments.

alan-agius4 reviewed (approved):
LGTM

josephperrott reviewed (approved):
LGTM

devversion commented on package.json:
```diff
@@ -24,6 +24,7 @@
     "/ ": "",
     "postinstall": "node scripts/webdriver-manager-update.js && node --preserve-symlinks --preserve-symlinks-main ./tools/postinstall-patches.js",
     "prepare": "husky install",
+    "ng-dev": "ts-node --esm --project .ng-dev/tsconfig.json node_modules/@angular/dev-infra-private/ng-dev/bundles/cli.mjs",
```
no, this will need to stay for longer. `ng-dev` is now a plain JS ESM tool, but it cannot load the `.ng-dev/*.mts` configuration files without registering `ts-node`. This is why we need to run `ng-dev` using `ts-node` here.

You may wonder why this worked without it before: The reason is that CJS ts-node loaders can be registered at runtime within Node, but ESM ts-node loaders cannot and need to be configured as part of starting the node runtime. e.g. when using `node --loaders my-typescript-esm-loader ng-dev/../cli.mjs`
